pkg/apis/core/validation/validation.go

```diff
@@ -1966,18 +1966,49 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *core.PersistentVolumeCl
 	if utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes) {
 		// lets make sure storage values are same.
 		if newPvc.Status.Phase == core.ClaimBound && newPvcClone.Spec.Resources.Requests != nil {
-			newPvcClone.Spec.Resources.Requests["storage"] = oldPvc.Spec.Resources.Requests["storage"]
+			newPvcClone.Spec.Resources.Requests[core.ResourceStorage] = oldPvc.Spec.Resources.Requests[core.ResourceStorage]
+		}
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.RecoverExpansionFailure) {
+			// set new PVC's allocatedResources to be old pvc allocatedResources so as
+			// even if they are nil validations could pass
+			newPvcClone.Spec.AllocatedResources = oldPvc.Spec.AllocatedResources
 		}
 
 		oldSize := oldPvc.Spec.Resources.Requests["storage"]
 		newSize := newPvc.Spec.Resources.Requests["storage"]
 
+		statusSize := oldPvc.Status.Capacity["storage"]
+
 		if !apiequality.Semantic.DeepEqual(newPvcClone.Spec, oldPvcClone.Spec) {
 			specDiff := diff.ObjectDiff(newPvcClone.Spec, oldPvcClone.Spec)
 			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), fmt.Sprintf("spec is immutable after creation except resources.requests for bound claims\n%v", specDiff)))
 		}
+
+		var oldAllocatedSize resource.Quantity
+		var newAllocatedSize resource.Quantity
+
+		if oldPvc.Spec.AllocatedResources != nil {
+			oldAllocatedSize = oldPvc.Spec.AllocatedResources.Requests[core.ResourceStorage]
+		}
+		if newPvc.Spec.AllocatedResources != nil {
+			newAllocatedSize = newPvc.Spec.AllocatedResources.Requests[core.ResourceStorage]
+		}
+
+		// allocatedResources is only allowed to grow. This check is not inside feature gate because it should
+		// not matter. Even if feature gate is disabled, we do not expect this invariant to change.
+		if newAllocatedSize.Cmp(oldAllocatedSize) < 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "allocatedResources", "requests", "storage"), "field can not be less than previous value"))
+		}
+
 		if newSize.Cmp(oldSize) < 0 {
-			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "resources", "requests", "storage"), "field can not be less than previous value"))
+			permitShrink := utilfeature.DefaultFeatureGate.Enabled(features.RecoverExpansionFailure)
+			if permitShrink && newSize.Cmp(statusSize) < 0 {
+				permitShrink = false
```

Comment:
Should the message be different in this case? The field **can** be less than previous value, it can't be less that PV size.